lib/util/paging-iterator.js

```diff
@@ -0,0 +1,190 @@
+	 * @constructor
+	 * @param {Object} response - The original API response
+	 * @param {BoxClient} client - An API client to make further requests
+	 * @returns {void}
+	 * @throws {Error} Will throw when collection cannot be paged
+	 */
+	constructor(response, client) {
+
+		var data = response.body;
+
+		if (!Array.isArray(data.entries)) {
+			throw new Error('Cannot create paging iterator for non-paged response!');
+		}
+
+		if (Number.isSafeInteger(data.offset)) {
+			this.nextField = PAGING_MODES.OFFSET;
+			this.nextValue = data.offset;
+		} else if (typeof data.next_marker !== 'undefined') {
+			this.nextField = PAGING_MODES.MARKER;
+			this.nextValue = data.next_marker;
+		} else {
+			throw new Error('Unable to determine paging strategy for response!');
+		}
+
+		this.limit = data.limit || data.entries.length;
+		this.done = false;
+
+		var queryIndex = response.request.href.indexOf('?');
+		var href = queryIndex > 0 ? response.request.href.substring(0, queryIndex) : response.request.href;
+		this.options = {
+			headers: response.request.headers,
+			qs: querystring.parse(response.request.uri.query)
+		};
+
+		// qs.parse() makes everything a string, ensure the limit is numeric
+		if (this.options.qs.limit) {
+			this.options.qs.limit = parseInt(this.options.qs.limit, 10);
+		}
+
+		delete this.options.headers.Authorization;
+		this.fetch = client.get.bind(client, href);
+		this.nextChunk = response.body.entries;
+		this.queue = new PromiseQueue(1, Infinity);
+		this._updatePaging(response);
+	}
+
+	/**
+	 * Update the paging parameters for the iterator
+	 * @private
+	 * @param {Object} response - The latest API response
+	 * @returns {void}
+	 */
+	_updatePaging(response) {
+
+		var data = response.body;
+
+		if (this.nextField === PAGING_MODES.OFFSET) {
+
+			this.nextValue += this.limit;
+
+			if (Number.isSafeInteger(data.total_count)) {
+				this.done = data.offset + this.limit >= data.total_count;
+			} else {
+				this.done = data.entries.length === 0;
+			}
+		} else if (this.nextField === PAGING_MODES.MARKER) {
+
+			if (data.next_marker) {
+				this.nextValue = data.next_marker;
+			} else {
+				this.nextValue = null;
+				this.done = true;
+			}
+		}
+
+		this.options.qs[this.nextField] = this.nextValue;
+	}
+
+	/**
+	 * Fetch the next page of results
+	 * @returns {Promise} Promise resolving to iterator state
+	 */
+	_getData() {
+
+		return Promise.fromCallback(callback => this.fetch(this.options, (err, response) => {
+
+			if (err) {
+				callback(err);
+				return;
+			}
+
+			if (response.statusCode !== 200) {
+				callback(errors.buildUnexpectedResponseError(response));
+				return;
+			}
+
+			this._updatePaging(response);
+
+			var result = {
+				done: this.done,
+				value: response.body.entries
+			};
+
+			callback(null, result);
+		}));
+	}
+
+	/**
+	 * Fetch the next page of the collection
+	 * @returns {Promise} Promise resolving to iterator state
+	 */
+	next() {
+
+		if (this.nextChunk) {
```

Comment:
I'm a little confused about this.nextChunk since you only use it here and in the constructor. Is it only for the initial results?